io/io/src/TDirectoryFile.cxx

```diff
@@ -1209,7 +1209,10 @@ void TDirectoryFile::ls(Option_t *option) const
       while (lnk) {
          TKey *key = (TKey*)lnk->GetObject();
          TString s = key->GetName();
-         if (s.Index(re) == kNPOS) continue;
+         if (s.Index(re) == kNPOS) {
+            lnk = lnk->Next();
```

Comment:
I think we can change [line 1209](https://github.com/root-project/root/pull/13217/files#diff-56ce80e060312dead2b5c7965a71f7ddee4211abf448bee96240b3c908833e7eR1209) to a `for` loop instead, as in
```c++
for (TObjLink *lnk = fKeys->FirstLink(); lnk != nullptr; lnk = lnk->Next()) {
```

That would result in keeping this line as-is and the removal of lines 1208 and 1225.  That should also result in smaller and more readable code.

Reply:
Done, good idea, thanks.

How should I address the other problem? A Slash within the key name prevents the object from being printed via ls(), for example 

```
TFile* aFile = new TFile("/tmp/tnamed.root", "RECREATE");
TNamed("TIME/CLK", "0x6").Write();
aFile->ls();
```

Works fine via TBrowser, object is shown with name, but ls() prints nothing in the terminal.

Reply:
> Done, good idea, thanks.
> 
> How should I address the other problem? A Slash within the key name prevents the object from being printed via ls(), for example
> 
> ```
> TFile* aFile = new TFile("/tmp/tnamed.root", "RECREATE");
> TNamed("TIME/CLK", "0x6").Write();
> aFile->ls();
> ```
> 
> Works fine via TBrowser, object is shown with name, but ls() prints nothing in the terminal.

@pcanal any suggestions?  That can be part of a follow-up PR anyways.  Many thanks for addressing this, @ferdymercury!

Reply:
This PR looks good as is.  Indeed we should address the issue in a subsequent PR.   I suspect the issue is that 'x/y' does not match the wildcard expression `*` (I guess it would/should match `*/*`).  The solution is likely to switch 
```
if (s.Index(re) == kNPOS) continue;
```
into
```
if (re_Is_set_by_the _user && s.Index(re) == kNPOS)
   continue;
```

Reply:
Sounds good, thanks for the hint. I will address this after this PR is merged. Cheers

Reply:
Matching `*/*` would then fail for other leafs that do not have the slash.

Here a reproducer for the ROOT terminal:

```
TString reg, s;
reg = "*"; s = "/"; s.Index(TRegexp(reg,kTRUE))
reg = "*"; s = "a"; s.Index(TRegexp(reg,kTRUE))
reg = "*/*"; s = "a"; s.Index(TRegexp(reg,kTRUE))
reg = "*/*"; s = "a/b"; s.Index(TRegexp(reg,kTRUE))
```

Reply:
> Side note: Matching */* would then fail for other leafs that do not have the slash.

Yes, it would have to be additive (and would 'technically' require an infinite number of wildcard expression to include all possible number of slashes.  Hence the recommendation of noting that the user did not pass a wildcard expression and that instead we should be matching "all" (without using any wildcard expression)